Source/BoundaryConditions/WarpX_PEC.cpp

```diff
@@ -686,82 +729,147 @@ PEC::ApplyReflectiveBoundarytoRhofield (
+
+        // Set psign on lo side
+        psign[idim][0] = ( (particle_boundary_lo[idim] == ParticleBoundaryType::Reflecting)
+                       ||  (particle_boundary_lo[idim] == ParticleBoundaryType::Thermal)
+                       ||  (field_boundary_lo[idim] == FieldBoundaryType::PMC) )
+                           ? 1._rt : -1._rt;
+
+        // Set psign on hi side
+        psign[idim][1] = ( (particle_boundary_hi[idim] == ParticleBoundaryType::Reflecting)
+                       ||  (particle_boundary_hi[idim] == ParticleBoundaryType::Thermal)
+                       ||  (field_boundary_hi[idim] == FieldBoundaryType::PMC) )
+                           ? 1._rt : -1._rt;
+
+        // Set the mirror index offset on lo and hi sides
         mirrorfac[idim][0] = 2*domain_lo[idim] - (1 - rho_nodal[idim]);
-        mirrorfac[idim][1] = 2*domain_hi[idim] + (1 - rho_nodal[idim]);
+        mirrorfac[idim][1] = 2*domain_hi[idim] - (1 - rho_nodal[idim]);
+
     }
-    const int nComp = rho->nComp();
 
 #ifdef AMREX_USE_OMP
 #pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
-    for (amrex::MFIter mfi(*rho); mfi.isValid(); ++mfi) {
+    for (amrex::MFIter mfi(*rho, false); mfi.isValid(); ++mfi) {
 
         // Get the multifab box including ghost cells
-        Box const& fabbox = mfi.fabbox();
+        const amrex::Box& rho_fabbox = mfi.fabbox();
 
-        // If grown_domain_box contains fabbox it means there are no PEC
-        // boundaries to handle so continue to next box
-        if (grown_domain_box.contains(fabbox)) { continue; }
+        // Get nodal box that does not include ghost cells
+        const amrex::Box node_box = amrex::convert(mfi.validbox(),IntVect::TheNodeVector());
 
-        // Extract field data
-        auto const& rho_array = rho->array(mfi);
+        //
+        // Step 1: Reflect Rho deposited to guard cells back into the domain
+        //
+        for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
 
-        // Loop over valid cells (i.e. cells inside the domain)
-        amrex::ParallelFor(mfi.validbox(), nComp,
-        [=] AMREX_GPU_DEVICE (int i, int j, int k, int n) {
-            amrex::ignore_unused(j,k);
-            // Store the array index
-            const amrex::IntVect iv(AMREX_D_DECL(i,j,k));
+            if ( !is_reflective[idim][0] && !is_reflective[idim][1] ) { continue; }
+
+            if ( (node_box.smallEnd()[idim] != domain_lo[idim]) &&
+                 (node_box.bigEnd()[idim] != domain_hi[idim]) ) { continue; }
+
+            // Get Rho box and grow to include guard cells in directions transverse
+            // to this boundary. This is required to correctly reflect Rho at domain
+            // corners that touch multiple PEC/PMC boundaries.
+            amrex::Box rho_box = amrex::convert(mfi.validbox(),rho_nodal);
+            for (int jdim = 0; jdim < AMREX_SPACEDIM; ++jdim) {
```

Comment:
I think a bit of work could be saved here by changing this to
`for (int jdim = idim+1; jdim < AMREX_SPACEDIM; ++jdim) {`
And the `if (jdim==idim) { continue; }` would not be needed.

Reply:
When idim == 1, I don't think the box needs to grow in dim 0, since that would update values in the guard cells that would then be overwritten in the call to ReflectJorRho. This is assuming that both routines would always be called together.